### PR TITLE
[Merged by Bors] - feat: product of sublattices

### PR DESCRIPTION
**Pull request by YaelDillies**
From LeanCamCombi

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/Sublattice.lean [hunk 1]
```diff
@@ -289,3 +289,90 @@ lemma map_top (f : LatticeHom α β) (h : Surjective f) : Sublattice.map f ⊤ =
...
+variable {L M : Sublattice α} {f : LatticeHom α β} {s t : Set α} {a : α}
+
+/-- Binary product of sublattices as a sublattice. -/
+def prod (L : Sublattice α) (M : Sublattice β) : Sublattice (α × β) where
+  carrier := L ×ˢ M
+  supClosed' := L.supClosed.prod M.supClosed
+  infClosed' := L.infClosed.prod M.infClosed
+
+@[simp, norm_cast] lemma coe_prod (L : Sublattice α) (M : Sublattice β) :
+    L.prod M = (L ×ˢ M : Set (α × β)) := rfl
```
Comment (b-mehta) [inline Mathlib/Order/Sublattice.lean:303]: surely simps can be used to make this one?

File: Mathlib/Order/Sublattice.lean [hunk 2]
```diff
@@ -289,3 +289,90 @@ lemma map_top (f : LatticeHom α β) (h : Surjective f) : Sublattice.map f ⊤ =
...
+
+@[simp] lemma mem_prod {M : Sublattice β} {p : α × β} : p ∈ L.prod M ↔ p.1 ∈ L ∧ p.2 ∈ M := Iff.rfl
+
+lemma prod_mono {L₁ L₂ : Sublattice α} {M₁ M₂ : Sublattice β} (hL : L₁ ≤ L₂) (hM : M₁ ≤ M₂) :
+    L₁.prod M₁ ≤ L₂.prod M₂ := Set.prod_mono hL hM
+
+lemma prod_mono_right {M₁ M₂ : Sublattice β} (hM : M₁ ≤ M₂) : L.prod M₁ ≤ L.prod M₂ :=
```
Comment (b-mehta) [inline Mathlib/Order/Sublattice.lean:310]: Let's state this in terms of `Monotone` too (or instead?)
Same below

File: Mathlib/Order/Sublattice.lean [hunk 3]
```diff
@@ -289,3 +289,90 @@ lemma map_top (f : LatticeHom α β) (h : Surjective f) : Sublattice.map f ⊤ =
...
+  prod_mono le_rfl hM
+
+lemma prod_mono_left {L₁ L₂ : Sublattice α} {M : Sublattice β} (hL : L₁ ≤ L₂) :
+    L₁.prod M ≤ L₂.prod M := prod_mono hL le_rfl
+
+lemma prod_top (L : Sublattice α) : L.prod (⊤ : Sublattice β) = L.comap LatticeHom.fst :=
```
Comment (b-mehta) [inline Mathlib/Order/Sublattice.lean:316]: Do you think this should be simp?

File: Mathlib/Order/Sublattice.lean [hunk 4]
```diff
@@ -289,3 +289,90 @@ lemma map_top (f : LatticeHom α β) (h : Surjective f) : Sublattice.map f ⊤ =
...
+  simp [SetLike.le_def, forall_and]
+
+@[simp] lemma prod_eq_bot {M : Sublattice β} : L.prod M = ⊥ ↔ L = ⊥ ∨ M = ⊥ := by
+  simpa only [← coe_inj] using Set.prod_eq_empty_iff
+
+@[simp] lemma prod_eq_top [Nonempty α] [Nonempty β] {M : Sublattice β} :
+    L.prod M = ⊤ ↔ L = ⊤ ∧ M = ⊤ := by simpa only [← coe_inj] using Set.prod_eq_univ
+
+/-- The product of sublattices is isomorphic to their product as lattices. -/
+def prodEquiv (L : Sublattice α) (M : Sublattice β) : L.prod M ≃o L × M where
```
Comment (b-mehta) [inline Mathlib/Order/Sublattice.lean:342]:
```suggestion
@[simps]
def prodEquiv (L : Sublattice α) (M : Sublattice β) : L.prod M ≃o L × M where
```

File: Mathlib/Order/Sublattice.lean [hunk 5]
```diff
@@ -289,3 +289,90 @@ lemma map_top (f : LatticeHom α β) (h : Surjective f) : Sublattice.map f ⊤ =
...
+
+/-- Arbitrary product of sublattices. Given an index set `s` and a family of sublattices
+`L : Π i, Sublattice (α i)`, `pi s L` is the sublattice of dependent functions `f : Π i, α i` such
+that `f i` belongs to `L i` whenever `i ∈ s`. -/
+def pi (s : Set κ) (L : ∀ i, Sublattice (π i)) : Sublattice (∀ i, π i) where
+  carrier := s.pi fun i ↦ L i
+  supClosed' := supClosed_pi fun i _ ↦ (L i).supClosed
+  infClosed' := infClosed_pi fun i _ ↦ (L i).infClosed
+
+@[simp, norm_cast] lemma coe_pi (s : Set κ) (L : ∀ i, Sublattice (π i)) :
```
Comment (b-mehta) [inline Mathlib/Order/Sublattice.lean:357]: as above

**Review by b-mehta — APPROVED** [review]

File: Mathlib/Order/Sublattice.lean (hunk 3 above)
Comment (YaelDillies) [reply Mathlib/Order/Sublattice.lean:316]: Not so sure, which is why I didn't tag it

Comment (b-mehta) [reply Mathlib/Order/Sublattice.lean:316]: Why not?

File: Mathlib/Order/Sublattice.lean [hunk 6]
```diff
@@ -289,3 +292,97 @@ lemma map_top (f : LatticeHom α β) (h : Surjective f) : Sublattice.map f ⊤ =
...
+@[simp] lemma mem_pi {s : Set κ} {L : ∀ i, Sublattice (π i)} {x : ∀ i, π i} :
+    x ∈ pi s L ↔ ∀ i, i ∈ s → x i ∈ L i := Iff.rfl
+
+@[simp] lemma pi_empty (L : ∀ i, Sublattice (π i)) : pi ∅ L = ⊤ := ext fun a ↦ by simp [mem_pi]
+
+@[simp] lemma pi_top (s : Set κ) : (pi s fun i ↦ ⊤ : Sublattice (∀ i, π i)) = ⊤ :=
+  ext fun a ↦ by simp [mem_pi]
+
+@[simp] lemma pi_bot [Nonempty κ] : (pi univ fun i ↦ ⊥ : Sublattice (∀ i, π i)) = ⊥ :=
+  ext fun a ↦ by simp [mem_pi]
```
Comment (b-mehta) [inline Mathlib/Order/Sublattice.lean:379]:
```suggestion
@[simp] lemma pi_bot {s : Set κ} (hs : s.Nonempty) :
    (pi s fun i ↦ ⊥ : Sublattice (∀ i, π i)) = ⊥ :=
  ext fun a ↦ by simpa [mem_pi] using hs

lemma pi_univ_bot [Nonempty κ] : (pi univ fun i ↦ ⊥ : Sublattice (∀ i, π i)) = ⊥ := by simp
```

File: Mathlib/Order/Sublattice.lean [hunk 7]
```diff
@@ -289,3 +292,97 @@ lemma map_top (f : LatticeHom α β) (h : Surjective f) : Sublattice.map f ⊤ =
...
+
+lemma le_pi {s : Set κ} {L : ∀ i, Sublattice (π i)} {M : Sublattice (∀ i, π i)} :
+    M ≤ pi s L ↔ ∀ i ∈ s, M ≤ comap (Pi.evalLatticeHom i) (L i) := by simp [SetLike.le_def]; aesop
+
+@[simp] lemma pi_univ_eq_bot {L : ∀ i, Sublattice (π i)} : pi univ L = ⊥ ↔ ∃ i, L i = ⊥ := by
```
Comment (b-mehta) [inline Mathlib/Order/Sublattice.lean:384]: Let's be consistent with `univ_pi_eq_empty_iff` here.

Comment (YaelDillies) [reply Mathlib/Order/Sublattice.lean]: It's not clear whether it should be `univ_pi_` or `pi_univ_` so I haven't changed that

File: Mathlib/Order/Sublattice.lean [hunk 8]
```diff
@@ -289,3 +292,101 @@ lemma map_top (f : LatticeHom α β) (h : Surjective f) : Sublattice.map f ⊤ =
...
+
+lemma le_pi {s : Set κ} {L : ∀ i, Sublattice (π i)} {M : Sublattice (∀ i, π i)} :
+    M ≤ pi s L ↔ ∀ i ∈ s, M ≤ comap (Pi.evalLatticeHom i) (L i) := by simp [SetLike.le_def]; aesop
+
+@[simp] lemma pi_univ_eq_bot_iff {L : ∀ i, Sublattice (π i)} : pi univ L = ⊥ ↔ ∃ i, L i = ⊥ := by
+  simp_rw [← coe_inj]; simp
+
+lemma pi_univ_eq_bot {L : ∀ i, Sublattice (π i)} {i : κ} (hL : L i = ⊥) : pi univ L = ⊥ :=
+  pi_univ_eq_bot_iff.2 ⟨i, hL⟩
+end Pi
```
Comment (b-mehta) [inline Mathlib/Order/Sublattice.lean:391]:
```suggestion
  pi_univ_eq_bot_iff.2 ⟨i, hL⟩

end Pi
```

**Comment by b-mehta** [comment]
Thanks!
bors merge

**Pull request closed without merging**